Docker file changes, shell script changes

Jeducious opened the pull request:
# Some Big Changes!

## Motivation
The main motivation here is to make the docker system more flexible, this means allowing users to build images themselves using whatever base image and version of Blender they wish. It also means allowing for using development branches of the addon inside a running container via a bind mount. The really helps include docker images in our development cycle as currently the only way to test if new code in the addon works in a container is to publish the addon live, and then pull it into a running container. 

## Changes

### Dockerfile
- Make it possible to build a functioning Blender/Crowdrender image from a base image like debian/Ubuntu/Nvidia etc. Optinos are available to choose the versino of Blender to install into the image.
- Removed the 'local' option, this is replaced by passing through the following options as env vars to the
   running container; persistent; token; secret; machine_uuid

### Shell scripts
 - Support the new --override-machine-uuid option for crowdrender.
 - Allow the addon to be mounted from a bind mount so that a development
    branch of Crowdrender can be used inside a container. This is via a new shell script run_dev_env.sh
 - Modified the install_addon script so that its possible ot run crowdrender from a bind mount
   instead of downloading a version from discovery.

### Other Chnages
 - Added a dockerignore file
 - Updated the make file to support docker building with other veresions of Blender.
 - Updated readme
 - Updated the github actions file to include options for building images with specified
    versions of blender.
Separately the secrets for the docket name and token have been changed on github to point to the new crowdrender docker hub.

Jeducious committed:
Docker file changes, shell script changes

Dockerfile
- Make it possible to build a functioning Blender/Crowdrender image from a base image like debian/Ubuntu/Nvidia etc. Optinos are available to choose the versino of Blender to install into the image.
- Removed the 'local' option, this is replaced by passing through the following options as env vars to the
   running container; persistent; token; secret; machine_uuid

Shell scripts
 - Support the new --override-machine-uuid option for crowdrender.
 - Allow the addon to be mounted from a bind mount so that a development
    branch of Crowdrender can be used inside a container.
 - Modified the install_addon script so that its possible ot run crowdrender from a bind mount
   instead of downloading a version from discovery.

Other Chnages
 - Added a dockerignore file
 - Updated the make file to support docker building with other veresions of Blender.
 - Updated readme
 - Updated the github actions file to include options for building images with specified
    versions of blender.
Separately the secrets for the docket name and token have been changed on github to point to the new crowdrender docker hub.

Jeducious committed:
Docker file changes, shell script changes

Dockerfile
- Make it possible to build a functioning Blender/Crowdrender image from a base image like debian/Ubuntu/Nvidia etc. Optinos are available to choose the versino of Blender to install into the image.
- Removed the 'local' option, this is replaced by passing through the following options as env vars to the
   running container; persistent; token; secret; machine_uuid

Shell scripts
 - Support the new --override-machine-uuid option for crowdrender.
 - Allow the addon to be mounted from a bind mount so that a development
    branch of Crowdrender can be used inside a container.
 - Modified the install_addon script so that its possible ot run crowdrender from a bind mount
   instead of downloading a version from discovery.

Other Chnages
 - Added a dockerignore file
 - Updated the make file to support docker building with other veresions of Blender.
 - Updated readme
 - Updated the github actions file to include options for building images with specified
    versions of blender.
Separately the secrets for the docket name and token have been changed on github to point to the new crowdrender docker hub.

zocker-160 commented on Dockerfile:
```diff
@@ -1,12 +1,42 @@
+ENV cr_version latest
+ENV persistent "false"
+ENV BL_VERSION_SHORT ${BL_VERSION_SHORT}
 
```
this is needed to make Nvidia GPUs work for the `--gpus` switch
you can always add it, because it does not harm non Nvidia / non GPU users

```suggestion

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES compute,utility

```

Jeducious replied:
Done, see diffs

zocker-160 commented on Dockerfile:
```diff
@@ -1,12 +1,42 @@
+ARG BL_VERSION_SHORT="3.2"
+ARG BL_VERSION_FULL="3.2.1"
+ARG BLENDER_BZ2_URL=https://download.blender.org/release/Blender${BL_VERSION_SHORT}/blender-${BL_VERSION_FULL}-${BLENDER_OS}.tar.xz
+ARG LOCAL_CR_PACKAGE=""
```
where is `LOCAL_CR_PACKAGE` used?

Jeducious replied:
It isn't so its removed now.

zocker-160 commented on Makefile:
```diff
@@ -1,4 +1,5 @@
+	docker build . --file Dockerfile --tag cr:cpu --build-arg BL_VERSION_SHORT="${BLENDER_VERSION_STRING}"
+
 gpu:
-	docker build . --file Dockerfile.cuda10 --no-cache --tag cr:gpu
+	docker build . --file Dockerfile.cuda10 --tag cr:gpu
```
I know this is a relic from my Makefile, but `Dockerfile.cuda10` does not exist.

Jeducious replied:
No worries mate ;) I just removed it and replaced this with a new docker file. Also, been a while hey! Took me some time to get around to finally doing this. Needed a reason and that reason may have just turned up.

zocker-160 replied:
haha yes, I think overall this would need a bit more of a rework, because there is no need for CUDA specific containers anymore as long as ppl run an up to date version of docker and nvidia-docker2.

Also I think in order to minimize maintenance, one could generate Dockerfiles on demand, IIRC that was a long term goal anyway.

zocker-160 commented on scripts/install_addon.py:
```diff
@@ -1,9 +1,27 @@
+# add arguments for installing and enabling the addon
+parser = optparse.OptionParser()
+# add option to install an addon from a file
+parser.add_option('-i', "--install", action="store_true", dest="install")
+parser.add_option("-e", "--enable", action="store_false", dest="install")
```
nitpick:
The `-e` option does not really make sense, because if conflicts with the `-i` switch and `addon_enable` gets called regardless.
I would rather add `default=False` to the `-i` option.
```suggestion
parser.add_option('-i', "--install", action="store_true", dest="install", default=False)
```

Jeducious replied:
Agree, not sure where that was going, esp a year on. Removed the -e enable option

zocker-160 commented on scripts/run_dev_env.sh:
```diff
@@ -0,0 +1,30 @@
+if [ -z "$docker_tag" ];
+then
+    
+    image=cr
```
nitpick:
whitespace :P
```suggestion
    image=cr
```

zocker-160 commented on Dockerfile:
```diff
@@ -1,12 +1,42 @@
+
+# Set environment vars to be used when the image is running in a container
+
+ENV local_cr_path "true"
```
true by default?

Jeducious replied:
false by default now.

zocker-160 commented on scripts/run_dev_env.sh:
```diff
@@ -0,0 +1,30 @@
+echo "Using $docker_tag tag'"
+
+
+docker run -dt --name "Crowdwrender-Server-DEV" -e token="$cr_token" --publish 9669-9694:9669-9694 --mount type=bind,source=$dev_location,destination="/root/.config/blender/$BL_VERSION_SHORT/scripts/addons/crowdrender" $image 
```
If I understood this correctly, this only works, because you set `local_cr_path` to `true` by default in `Dockerfile`.

IMO it would make more sense to disable it by default and then add `-e local_cr_path=true` in this line here.

Jeducious replied:
local_cr_path is no longer set true by default now.

zocker-160 commented on scripts/start_cr_server.sh:
```diff
@@ -2,76 +2,116 @@
+    
+    # check to see if the local dev path is set, don't install the
+    # addon from discovery if so
+    if ! [ -z "$local_cr_path" ];
```
by default this is always true, isn't it?

Jeducious replied:
Not anymore, the user must specify it as true and provide a bind mount to supply the locally hosted addon. This is mainly a development feature now, not used in normal operations.

zocker-160 commented on .github/workflows/dockerimage-latest.yml:
```diff
@@ -38,10 +37,8 @@ jobs:
-        docker build . --pull --no-cache --file Dockerfile.293LTS --tag crowdrender/blender-crowdrender:bl_2.93
-        
-        docker push crowdrender/blender-crowdrender:bl_2.93
+        docker build . --pull --no-cache --file Dockerfile.cuda10 --tag crowdrender/blender-crowdrender:bl_2.83_cu_10
```
I know this is a relic of mine, but this will fail because `Dockerfile.cuda10` does not exist.

Jeducious replied:
I made a lot of changes here to support building multiple images. I've removed the reference to cuda10.

zocker-160 commented on .github/workflows/dockerimage-latest.yml:
```diff
@@ -54,7 +51,7 @@ jobs:
-        docker build . --pull --no-cache --file Dockerfile.283LTS --tag crowdrender/blender-crowdrender:bl_2.83
-        
-        docker push crowdrender/blender-crowdrender:bl_2.83
+        docker build . --pull --no-cache --file Dockerfile.cuda11 --tag crowdrender/blender-crowdrender:nvidia
```
same as above

Jeducious replied:
resolved as above

zocker-160 commented on scripts/start_cr_server.sh:
```diff
@@ -2,76 +2,116 @@
     else
-        echo "downloading CR version: $CR_VERSION!"
-        curl -# -H "Authorization: Bearer $token" -X GET https://discovery.crowd-render.com/api/download/addon/$CR_VERSION.zip -o $CR_VERSION.zip
+        # 
```
a lonely hashtag :P

Jeducious replied:
lonely hash's day got worse, I deleted it.

zocker-160 commented on scripts/start_cr_server.sh:
```diff
@@ -2,76 +2,116 @@
+    if ! [ -z "$local_cr_path" ];
+    then
+        echo "using mounted location for CR addon"
+        /usr/local/blender/blender -b -noaudio --python ./install_addon.py -- -e
```
in case you accept my suggestion in `install_addon.py`
```suggestion
        /usr/local/blender/blender -b -noaudio --python ./install_addon.py
```

Jeducious replied:
accepted!

zocker-160 commented on scripts/start_cr_server.sh:
```diff
@@ -2,76 +2,116 @@
+
+
+    # exit early if a docker bind mount is specified, but no version number is provided.
+    if [ -z "$local_cr_path" ] && [ -z "$cr_version" ]; then
```
I don't quite understand this. 

If I understood this correctly, this will only be true if someone manually sets `-e local_cr_path=""` which seems unlikely since default is `true`.

Is `local_cr_path` used inside crowdrender?

If not I think it would make more sense to set default to `false` and then check against `true`.

zocker-160 commented on scripts/start_cr_server.sh:
```diff
@@ -2,76 +2,116 @@
+    if ! [ -z "$local_cr_path" ]; then
+        
+        # use the provided path as the source of the addon package instead of downloading it
+        echo "Going to use $local_cr_path as the source of the addon package."
```
is it though?

Jeducious replied:
resolved as I've changed this function to reflect the fact that the use_local_cr is false by default.

zocker-160 commented on .github/workflows/dockerimage-latest.yml:
```diff
@@ -17,15 +17,14 @@ jobs:
-        
-        docker tag crowdrender/blender-crowdrender:latest crowdrender/blender-crowdrender:bl_3.2
-        
+        docker build . --pull --no-cache --file Dockerfile --tag crowdrender/blender-crowdrender:latest --build-arg BLENDER_BZ2_URL="$BLENDER_URL" BL_VERSION_SHORT="${BLENDER_VERSION_STRING}"
```
why set `BLENDER_BZ2_URL`? 

Setting `BL_VERSION_SHORT` and `BL_VERSION_FULL` would make more sense, given that you assemble the url in the Dockerfile using those ARGs.

zocker-160 replied:
you will get error here now because `BLENDER_VERSION_STRING` does not exist :V

Jeducious replied:
I've refactored a lot of this file now, could you take another look? The aim is to specify a set of versions to build and then push images based on those versions.

zocker-160 reviewed (changes requested):
Looks good overall, there are a few blockers and a lot of nitpicks.

Also it would be good if you rebased this on top of master in order to have the updated readme, since you still have the old instructions for Nvidia users in this branch.

Jeducious committed:
Blender Download URL changed

The Blender download URL was specified as a separate input to the github action file. This is changed so that now the URL is constructed from other variables, namely the short and long version string, allowing you to customise the version of Blender used.

I also removed a build arg from the make file as it was overriding the version number with nothing causing a build error.

Jeducious committed:
Blender Download URL changed

The Blender download URL was specified as a separate input to the github action file. This is changed so that now the URL is constructed from other variables, namely the short and long version string, allowing you to customise the version of Blender used.

I also removed a build arg from the make file as it was overriding the version number with nothing causing a build error.

Jeducious committed:
Update README.md

Jeducious committed:
Merge branch 'fix-machine-uuid-overriding' of https://github.com/crowdrender/cr-docker into fix-machine-uuid-overriding

Jeducious committed:
Add a new docker file based on CUDA 12

Jeducious committed:
Update make file

Removing refs to deprecated Dockerfiles, updating tags as well

Jeducious committed:
Removing LTS dockerfiles

Jeducious committed:
Renamed variables to match build args used in GH actions file

Jeducious committed:
Removed uncessary nvidia toolkit package

Tested on windows, running WSL2, I can render using an RTX GPU in a container. Seems like the container toolkit isn't required.

Jeducious committed:
Use matrix builds

Supporting the building of multiple Blender versions is now a lot easier.

I used the matrix builds in actions to create a list of blender versions that will be supported.

Some minor changes to the naming of the repository and the tags are in here too.

Jeducious committed:
removed uneeded options

Only need the filepath to the addon package now, other options have been removed.

Support for using a copy of the addon that is on the host, but not in the container is done by simply omitting the -i flag, and making sure the local copy is bound to the container using docker's bind mounts.

Jeducious committed:
Removed unecessary args

The sript called for both an image name and a tag, its simpler to just accept the image name with the tag included in it.

Removed checking for this tag as well.

Added a variable to hold the path ot the location the addon should be installed when running a dev session. All dev sessions use a copy of the addon hosted on the host, not in the container, hence the use of a bind mount.

Jeducious committed:
Update to support V0.6.4 CR addon

The new V0.6.4 and beyond will reequire a change to the commands used to start a server/worker.

Also, the var "use_local_cr" is now false by default and I've made some effort to consolidate and tidy up.

Jeducious committed:
Remove unecessary quote marks

Docker interprets ENV key value such that the value is always a string, no need for explicit quotation marks.

Jeducious committed:
Add nvidia-container-toolkit

Nvidia seems to have moved drivers into the toolkit, the cuda/driver images are depracated.

Needs testing on a system that has a GPU though.

Jeducious committed:
Update readme

Reflects changes to the tags, repo name, supported versions of Blender.

Two sections now, one for support of Docker Desktop and one for linux.

Jeducious committed:
Changed tag names

Better reflects the intended target system.

Jeducious commented:
Hi @zocker-160, I've just pushed commits to this branch/PR. Would really appreciate if you can let me know what you think and if we can merge. I really would like to get these containers back in use, since we've not updated them in ages, I think people are less likely to use them since Blender has moved on and is about to move to V4.0.

I think the changes I've made will help, I've tested GPU support on Docker Desktop in windows with a new docker file that is based on the original. All works well.

Cheers!

Jeducious committed:
Fix errors in GH actions

Should have supervised the AI better.

Jeducious committed:
Fixing a bug with GH actions

matrix vars not available to b used in the command to build the blender url resulted in a useless URL and no blender being downloaded.

Jeducious committed:
Configure repos for Nvidia container toolkit

Jeducious committed:
Adding repo's requires gnupg

this package isn't included in slim, so adding it.

Jeducious committed:
Fix bug with actions file

Was using matrix builds incorrectly, resulting in the Blender URL being wrong.

I've now paired up the short and full version names so that the builds use the correct short and full versions when constructing the blender download URL.

Jeducious committed:
Fixed issue with docker desktop blender url

Forgot to carry changes from previous commit to the same lines in the docker desktop build.

Jeducious committed:
Fixed bug in actions file

Wrong name for the image in the build command.

zocker-160 commented on Dockerfile:
This file will need some changes, as there are missing dependencies and a broken ENTRYPOINT. 
I will push those changes myself.

zocker-160 replied:
changes pushed

zocker-160 reviewed (approved):
Small changes needed to make base image run on native Linux machine.

zocker-160 committed:
add .bak to gitignore

zocker-160 committed:
update copyright year in "header"

zocker-160 committed:
fix missing dependencies and cleanup install process

zocker-160 committed:
replace curl command with docker ADD

zocker-160 committed:
fix wrong path in entrypoint

zocker-160 commented:
Seems to be working from my short testing, merging now after our conversation on Slack, but more changes will be needed (especially in the Readme which is kinda a mess IMO).
